File: javascript/src/utils/schema.ts
import { dedent } from 'ts-dedent'

import { InvalidArgumentsError } from './errors'
import { ArgumentType, Arguments, TransformedArguments } from '../types'
import { isFalsyArgumentValue, isBooleanString, isNumberString, isRecord, pick } from './helpers'

class ArgumentParser {
  private schema: Arguments

  constructor(schema: Arguments) {
    this.schema = schema
  }

  parse(argumentsToParse: Record<string, unknown>): {
    parsedArguments: Record<string, unknown>
    parsedErrors: string[]
  } {
    const parsedArguments: Record<string, unknown> = { ...argumentsToParse }
    const parsedErrors: string[] = []

    for (const argName in this.schema) {
      const argDefinition = this.schema[argName]
      const argValue = argumentsToParse[argName]
      const falsyArgValue = isFalsyArgumentValue(argValue)

      if (falsyArgValue) {
        if (argDefinition.required) {
          parsedErrors.push(`Argument "${argName}" is required but missing.`)
          continue
        }
      }

      if (!falsyArgValue) {
        if (argDefinition.array) {
          if (!Array.isArray(argValue)) {
            parsedErrors.push(`Argument "${argName}" should be an array.`)
            continue
          }

          const parseValueTypeItems = argValue.map((item) => this.parseValueType(item, argDefinition.type))
          if (!parseValueTypeItems.every((item) => item.isValidType)) {
            parsedErrors.push(`Argument "${argName}" should be an array of type "${argDefinition.type}".`)
          }

          parsedArguments[argName] = parseValueTypeItems.map((item) => item.parsedValue)
        } else {
          const parsedValueTypeItem = this.parseValueType(argValue, argDefinition.type)
          if (!parsedValueTypeItem.isValidType) {
            parsedErrors.push(`Argument "${argName}" should be of type "${argDefinition.type}".`)
          }

          parsedArguments[argName] = parsedValueTypeItem.parsedValue
        }
      }
    }

    return { parsedErrors, parsedArguments }
  }

  private parseValueType(
    value: unknown,
    expectedType: ArgumentType,
  ): {
    isValidType: boolean
    parsedValue: unknown
  } {
    switch (expectedType) {
      case 'string':
        // For string type, we just need to convert to string.
        return {
          isValidType: true,
          parsedValue: typeof value === 'object' ? JSON.stringify(value) : String(value), // Need to do this because String(object) returns "[object Object]".
        }
      case 'boolean':
        // For boolean type, we need to check if the value is a boolean or a boolean string.
        if (typeof value === 'boolean') {
          return {
            isValidType: true,
            parsedValue: value,
          }
        }

        if (isBooleanString(value)) {
          return {
            isValidType: true,
            parsedValue: value.toLowerCase() === 'true',
          }
        }

        return {
          isValidType: false,
          parsedValue: value,
        }
      case 'number':
        // For number type, we need to check if the value is a number or a number string.
        if (typeof value === 'number') {
          return {
            isValidType: true,
            parsedValue: value,
          }
        }

        if (isNumberString(value)) {
          return {
            isValidType: true,
            parsedValue: parseFloat(value),
          }
        }

        return {
          isValidType: false,
          parsedValue: value,
        }
      case 'dict':
        // For dict type, we need to check if the value is a record.
        if (isRecord(value)) {
          return {
            isValidType: true,
            parsedValue: value,
          }
        }

        return {
          isValidType: false,
          parsedValue: value,
        }
      case 'json':
        // For json type, we need to check if the value is a valid JSON string.
        try {
          const parsedJSONValue = JSON.parse(JSON.stringify(value))
          return {
            isValidType: true,
            parsedValue: parsedJSONValue,
          }
        } catch {
          return {
            isValidType: false,
            parsedValue: value,
          }
        }
      default:
        expectedType satisfies never
        throw new Error(`Unknown argument type "${expectedType}".`)
    }
  }
}

export function parseFunctionArguments(args: unknown, schema: Arguments): TransformedArguments<Arguments> {
  if (!isRecord(args)) {
    throw new Error(`The given arguments are invalid.`)
  }

  const argumentParser = new ArgumentParser(schema)
  const { parsedArguments, parsedErrors } = argumentParser.parse(args)

  if (parsedErrors.length > 0) {
    const invalidArgumentsError = dedent`
      Invalid parameter(s) found:
      ${parsedErrors.join('\n')}
    `

    throw new InvalidArgumentsError(invalidArgumentsError)
  }

  return pick(parsedArguments, Object.keys(schema)) as TransformedArguments<Arguments>
}

File: python/retoolrpc/utils/schema.py
from typing import Any, Dict, List, Tuple

from retoolrpc.utils.errors import InvalidArgumentsError
from retoolrpc.utils.helpers import (
    is_boolean_string,
    is_falsy_argument_value,
    is_json_value,
    is_number_string,
)
from retoolrpc.utils.types import Arguments, ArgumentType


class ArgumentParser:
    def __init__(self, schema: Arguments):
        self.schema = schema

    def parse(
        self, arguments_to_parse: Dict[str, Any]
    ) -> Tuple[Dict[str, Any], List[str]]:
        parsed_arguments = arguments_to_parse.copy()
        parsed_errors = []

        for arg_name, arg_definition in self.schema.items():
            arg_value = arguments_to_parse.get(arg_name)
            falsy_arg_value = is_falsy_argument_value(arg_value)

            if falsy_arg_value and arg_definition["required"]:
                parsed_errors.append(f'Argument "{arg_name}" is required but missing.')
                continue

            if not falsy_arg_value:
                if arg_definition["array"]:
                    if not isinstance(arg_value, list):
                        parsed_errors.append(
                            f'Argument "{arg_name}" should be an array.'
                        )
                        continue

                    parse_value_type_items = [
                        self.parse_value_type(item, arg_definition["type"])
                        for item in arg_value
                    ]
                    if not all(item[0] for item in parse_value_type_items):
                        parsed_errors.append(
                            f'Argument "{arg_name}" should be an array of type '
                            f'"{arg_definition["type"]}".'
                        )

                    parsed_arguments[arg_name] = [
                        item[1] for item in parse_value_type_items
                    ]
                else:
                    is_valid_type, parsed_value = self.parse_value_type(
                        arg_value, arg_definition["type"]
                    )
                    if not is_valid_type:
                        parsed_errors.append(
                            f'Argument "{arg_name}" should be of type '
                            f'"{arg_definition["type"]}".'
                        )

                    parsed_arguments[arg_name] = parsed_value

        return parsed_arguments, parsed_errors

    def parse_value_type(
        self, value: Any, expected_type: ArgumentType
    ) -> Tuple[bool, Any]:
        if expected_type == "string":
            return True, str(value)
        elif expected_type == "boolean":
            if isinstance(value, bool):
                return True, value
            elif is_boolean_string(value):
                return True, value.lower() == "true"
            else:
                return False, value
        elif expected_type == "number":
            if isinstance(value, (int, float)):
                return True, value
            elif is_number_string(value):
                return True, float(value)
            else:
                return False, value
        elif expected_type == "dict":
            if isinstance(value, dict):
                return True, value
            else:
                return False, value
        elif expected_type == "json":
            if is_json_value(value):
                return True, value
            else:
                return False, value
        else:
            raise ValueError(f"Unknown argument type '{expected_type}'.")


def parse_function_arguments(args: Any, schema: Arguments) -> Dict[str, Any]:
    if not isinstance(args, dict):
        raise ValueError("The given arguments are invalid.")

    argument_parser = ArgumentParser(schema)
    parsed_arguments, parsed_errors = argument_parser.parse(args)

    if parsed_errors:
        error_message = "\n".join(["Invalid parameter(s) found:"] + parsed_errors)
        raise InvalidArgumentsError(error_message)

    return {k: v for k, v in parsed_arguments.items() if k in schema.keys()}
